mk_server: introduce start event loop in lib mode

braydonk opened the pull request:
Closes #380 

This patch addresses a race condition in `mk_start` under `MK_SCHEDULER_FAIR_BALANCING` mode wherein the worker started by `mk_start` could accidentally receive the `MK_SERVER_SIGNAL_START` it is trying to send to the main thread. This patch introduces `mk_server->lib_evl_start`, which is a small event loop used exclusively by the `mk_start` thread to receive the start signal from the worker. This eliminates the race condition, since the worker and main thread will no longer be racing to receive the `MK_SERVER_SIGNAL_START` signal.

Signed-off-by: braydonk <braydonk@google.com>

edsiper commented:
assigning @leonardo-albertovich as reviewer

braydonk commented:
Used this [test program](https://github.com/braydonk/monkey_test_program) to build a barebones library server scenario. Here are the valgrind results running with this PR:
```bash
braydonk@bk:~/Git/monkey_test_program/build$ valgrind --leak-check=full -s ./mk 
==2113030== Memcheck, a memory error detector
==2113030== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==2113030== Using Valgrind-3.19.0 and LibVEX; rerun with -h for copyright info
==2113030== Command: ./mk
==2113030== 
starting the server
^Cwe're outta here
we're cleaning up
==2113030== 
==2113030== HEAP SUMMARY:
==2113030==     in use at exit: 0 bytes in 0 blocks
==2113030==   total heap usage: 76 allocs, 76 frees, 91,142 bytes allocated
==2113030== 
==2113030== All heap blocks were freed -- no leaks are possible
==2113030== 
==2113030== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```
With this I think I'm ready to take it out of draft state.

leonardo-albertovich commented on mk_server/monkey.c:
```diff
@@ -122,6 +122,27 @@ struct mk_server *mk_server_create()
+    /* Library mode: start event loop */
+    server->lib_evl_start = mk_event_loop_create(1);
+    if (!server->lib_evl_start) {
+        mk_mem_free(server);
```
This call is missing here `mk_event_loop_destroy(server->lib_evl);`

braydonk replied:
Nice catch thanks! Fixed in [efd10da](https://github.com/monkey/monkey/pull/381/commits/efd10dabdd2012dc6bff74691bb833ef62a16e2e)

leonardo-albertovich reviewed (approved):
I don't see any issues with the code and it seems to work (I tested it in a container). The only thing I'd like to get a clearer idea of is why are we using a second event loop just for the start channel. It seems like something that could be done in an simpler way but truth be told I'm not sure if I'd trade homogeneity / certainty for simplicity or resource saving at this particular point (especially considering that this isn't something that's performed repeatedly).

I would really like to add accutest to monkey and have your test program in a test case, if you did that it'd be great but otherwise it's great as is and a much welcome improvement from my point of view.

Thanks a lot.

braydonk commented:
Thanks for the review!

> The only thing I'd like to get a clearer idea of is why are we using a second event loop just for the start channel. It seems like something that could be done in an simpler way

I had 2 ideas to try and sync up the threads to avoid the race condition laid out in #380. My first idea was to add a start mutex that the main thread would lock and unlock once it started, and the worker threads wouldn't start until the mutex is available. I had trouble getting this working, though it's been a couple days and I don't remember exactly what the issue I was running into there was. I ended up going with this second idea of a small event loop that is destroyed after start, which was the next cheapest thing I could think of. If you think it's worth revisiting the mutex idea I can take another look and see if I can get that working.

> I would really like to add accutest to monkey and have your test program in a test case

I can give it a try!  

I would probably do something like:
1. Make new directory `test`
2. Add a `lib` folder inside `test` where I'd add `acutest` (alternatively, I could simply put the `acutest.h` header in there? depends how we'd like to manage that)
3. I'd create `lib_test.c` and add the test program as a test case there
4. Add a `CMakeLists.txt` that makes `lib_test` executable
5. Add an `MK_LIB_TESTS` option in the main `CMakeLists.txt` to build the tests.

Does this make sense? Is there any different way you'd prefer the tests be hooked up?

leonardo-albertovich commented:
If I thought of this in pthread terms I'd use a condition instead of a mutex but honestly, I wouldn't do either thing because we use a pthread emulation layer in windows and that would mean using more of it which could introduce bugs since I doubt anyone ever reviewed it more than going through it to debug something.

I was originally thinking of a socket pair and a timed recv. I know channels are implemented socketpairs as provided by libev in windows but are real pipes in linux which would break that approach (unless more code was added to explicitly use a socketpair in every platform to be able to set the recv timeout with setsockopt).

From a practical point of view I think what you did is correct, it might not be perfect but this codebase is far from perfect so I'm happy with the improvement and peace of mind that reusing a well known mechanism / pattern offers.

As for tests yes, I think what you described is correct. If you need a reference you can check cmetrics, ctraces or cfl, those are smallish libraries created by the same team that use accutest for the unit tests. If you get stuck at any point just ping me and I'll give you a hand.

braydonk commented:
Sounds great, thanks! I'll get started on adding acutest support. Is there a Slack channel or somewhere that I can reach you and/or the Monkey team directly? (I'm already in the Fluent slack if that works, but if there's somewhere more specific I can go there)

braydonk commented:
I've added an acutest setup and a test suite for server start and stop. It ended up needing more work to set up than I expected, however because the "failure case" so to speak of the server starting and stopping could be an infinite hangup, I decided to set it up as two workers where one is a timeout. Hopefully it's not too overkill and will be usable for more test cases in the future.

Here are the valgrind results for the test cases I added so far.

```
braydonk@bk:~/Git/monkey_test_program/monkey/build$ valgrind --leak-check=full -s ./bin/mk-test-lib_server 
==2290234== Memcheck, a memory error detector
==2290234== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==2290234== Using Valgrind-3.19.0 and LibVEX; rerun with -h for copyright info
==2290234== Command: ./bin/mk-test-lib_server
==2290234== 
Test server_start_stop...                       [ OK ]
==2290235== 
==2290235== HEAP SUMMARY:
==2290235==     in use at exit: 0 bytes in 0 blocks
==2290235==   total heap usage: 81 allocs, 81 frees, 70,378 bytes allocated
==2290235== 
==2290235== All heap blocks were freed -- no leaks are possible
==2290235== 
==2290235== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
Test server_start_stop_multi_worker...          [ OK ]
==2290242== 
==2290242== HEAP SUMMARY:
==2290242==     in use at exit: 0 bytes in 0 blocks
==2290242==   total heap usage: 235 allocs, 235 frees, 462,938 bytes allocated
==2290242== 
==2290242== All heap blocks were freed -- no leaks are possible
==2290242== 
==2290242== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
Test server_start_stop_force_fair_balancing...  [ OK ]
==2290255== 
==2290255== HEAP SUMMARY:
==2290255==     in use at exit: 0 bytes in 0 blocks
==2290255==   total heap usage: 85 allocs, 85 frees, 91,922 bytes allocated
==2290255== 
==2290255== All heap blocks were freed -- no leaks are possible
==2290255== 
==2290255== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
SUCCESS: All unit tests have passed.
==2290234== 
==2290234== HEAP SUMMARY:
==2290234==     in use at exit: 0 bytes in 0 blocks
==2290234==   total heap usage: 2 allocs, 2 frees, 1,072 bytes allocated
==2290234== 
==2290234== All heap blocks were freed -- no leaks are possible
==2290234== 
==2290234== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```
